Prevent some rebuilds for future Nix reformats

infinisil opened the pull request:
## Description of changes
This is motivated by https://github.com/NixOS/nixpkgs/pull/322537, which I'd like to get close to 0 rebuilds, although due to the nature of that PR not formatting files modified by other PRs, me making _this_ PR makes the other PR avoid these files entirely, so this PR doesn't have to be merged before the other one.

This PR either avoids Nix files ending up in the build result (so that changing them won't cause rebuilds), or formats Nix files in advance, so that they don't need to be changed in the treewide reformat.

This is effectively part of https://github.com/NixOS/nixpkgs/issues/301014

## Things done
- [x] Built all the changed derivations

---

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
lib/tests: Format a file with nixfmt

And fix locations to not break the test.
This is a rare case where another change is required after formatting.
We do this in a separate commit so that we don't need to do it in the
treewide reformatting PR.

infinisil committed:
nix-info: Format files with nixfmt

This is a rare case of a Nix file actually ending up in the build result.
We reformat this now, causing a rebuild, so that we won't cause a rebuild
in the treewide reformatting PR.

infinisil committed:
tests.trivial-builders: Prevent rebuilds when Nix files change

There's no need to use a Nix file in the path here. By using a different
file we won't cause rebuilds when we change the Nix file, in particular
also when the Nix file is reformatted.

infinisil committed:
tests.makeBinaryWrapper: Prevent rebuilds when Nix files change

The source included way more files than it really needed.
This commit limits it to exactly those it needs.

This also makes sure that no rebuild is necessary when any Nix
file changes, in particular useful when we reformat the Nix files.

infinisil committed:
vimPluginsUpdater: Format files with nixfmt

This is a rare case of a Nix file actually ending up in the build result.
We reformat this now, causing a rebuild, so that we won't cause a rebuild
in the treewide reformatting PR.

Atemu commented:
> or formats Nix files in advance, so that they don't need to be changed in the treewide reformat.

This doesn't strike me as a great solution tbh. It just kicks the bucket down the road and will cause rebuilds again when the formatting rules are changed again in the future.

sternenseemann commented:
The `cabalSdist` test can be made to not depend on the formatting of the *generated* Nix file by using `haskell.lib.compose.overrideSrc` at the call site to use a filtered source. If this filtered source is also passed correctly to the other derivation that uses it (near the call site of `generated.nix`), all should be fine.

See #320572, #321246.

infinisil committed:
tests.haskell.cabalSdist: Prevent rebuilds when Nix files change

The generated file sets its own directory as the source, including the
generated file itself, which causes rebuilds when that file is
reformatted. We can avoid this by overriding the source with a filtered
version and using that throughout the tests.

See https://github.com/NixOS/nixpkgs/pull/320572 for more context

infinisil commented:
> This doesn't strike me as a great solution tbh. It just kicks the bucket down the road and will cause rebuilds again when the formatting rules are changed again in the future.

Speaking for the formatting team, we don't expect to have to make any significant changes to the format in the foreseeable future. The only changes to expect soon are bug fixes, which won't change much. So I don't think we need to worry very much about it :)

@sternenseemann Thanks, done that in the last commit now!

philiptaron commented on pkgs/applications/editors/vim/plugins/get-plugins.nix:
Sigh, this is on my list of #208242 things to go make better.

philiptaron commented on pkgs/test/make-binary-wrapper/default.nix:
```diff
@@ -15,16 +15,25 @@ let
     mkdir -p tmp/foo # for the chdir test
 
-    params=$(<"${./.}/${testname}.cmdline")
+    source=${lib.fileset.toSource {
+      root = ./.;
+      fileset = lib.fileset.unions [
+        (./. + "/${testname}.cmdline")
+        (./. + "/${testname}.c")
+        (lib.fileset.maybeMissing (./. + "/${testname}.env"))
+      ];
+    }}
```
This would be clearer if extracted to another binding.

infinisil replied:
I think this is fine, I don't think much is gained from having an intermediate Nix binding (we already assign it to a bash variable here)

philiptaron commented on pkgs/test/haskell/cabalSdist/default.nix:
```diff
@@ -1,7 +1,18 @@
 
 let
-  localRaw = haskellPackages.callPackage ./local/generated.nix {};
+  src = lib.fileset.toSource {
+    root = ./local;
+    fileset = lib.fileset.unions [
+      ./local/app
+      ./local/CHANGELOG.md
+      ./local/local.cabal
+    ];
+  };
+  # This prevents the source from depending on the formatting of the ./local/generated.nix file
+  localRaw = haskell.lib.compose.overrideSrc {
+    inherit src;
+  } (haskellPackages.callPackage ./local/generated.nix {});
```
Having the `src = ./.` line in `local/generated.nix` is not needed now, right? It can be the same as the `src` above?

infinisil replied:
Since that's technically a generated file, I think it's cleaner to override the source from here instead

philiptaron replied:
It's named as such but is in no way generated.

infinisil replied:
I'm pretty sure it was generated using `cabal2nix`, e.g. check see [here](https://github.com/NixOS/cabal2nix/blob/92c93710247e6a4a827515caee778f8c8b3b24ea/cabal2nix/README.md?plain=1#L44-L56), but I guess that's not very relevant at this point anyways

philiptaron reviewed (commented):
A few curious comments.

0x4A6F commented:
Please also tag @NixOS/nix-formatting for PRs touching that topic.

roberth commented on pkgs/applications/editors/vim/plugins/get-plugins.nix:
```diff
@@ -1,19 +1,28 @@
+  hasChecksum =
+    value:
+    lib.isAttrs value
+    && lib.hasAttrByPath [
+      "src"
+      "outputHash"
+    ] value;
```
As a nixpkgs-fmt user this feels ugly.
I wonder if this would be ok? I think it's cleaner to use the built-in syntax anyway.

```suggestion
    && (value?src.outputHash);
```

infinisil replied:
Indeed this can be simplified and reformatted to just
```
  hasChecksum = value: value ? src.outputHash;
```

[Relevant motto](https://github.com/NixOS/rfcs/blob/62d1245ec1eae275edb996229585637035c02538/rfcs/0166-nix-formatting.md?plain=1#L78) from the RFC:

> *Bad code does not deserve good formatting.*

:laughing:

roberth commented on lib/tests/modules/declaration-positions.nix:
```diff
@@ -27,23 +25,25 @@ assert builtins.unsafeGetAttrPos "a" (discardPositions { a = true; }) == null;
-  options.nested.nestedLine30 = lib.mkOption {
-    type = lib.types.int;
-  };
+  options.nested.nestedLine30 = lib.mkOption { type = lib.types.int; };
```
I like my `mkOption`s multi-line, consistently.
Will I have to adjust?

infinisil replied:
Oh this is actually an inconsistency in the formatter that should be fixed, because
```nix
  options.nested.nestedLine30 = { type = lib.types.int; };
```

expands to
```nix
  options.nested.nestedLine30 = {
    type = lib.types.int;
  };
```

The same should be done here.

infinisil replied:
This is actually less prevalent in practice than it looks like here, because as soon as you have at least two attribute values, it will also get expanded. And effectively all options in NixOS have both a `type` and a `description`, which makes two.

roberth replied:
I figured it might not be as bad.

Sometimes I start out with only a `description` or only a `type` if it's more "technical" like testing or trying something out.
I do anticipate the other attrs, which is why this stood out to me.

roberth commented on pkgs/tools/nix/info/relaxedsandbox.nix:
```diff
@@ -1,12 +1,9 @@
-  ''
+  pkgs = import <nixpkgs> { };
+in
+pkgs.runCommand "diagnostics-sandbox" { __noChroot = true; } ''
```
Do we really need to collapse this attrset into a single line format?
This will add a bunch of diff noise when new attrset are added, won't it?

infinisil replied:
I'm not sure if we'll go for that in the end, but with https://github.com/NixOS/nixfmt/pull/222 (which also fixes your other complaint), this file would look like this instead:

```nix
let
  pkgs = import <nixpkgs> { };
in
pkgs.runCommand "diagnostics-sandbox"
  {
    __noChroot = true;
  }
  ''
    set -x
    # no cache: ${toString builtins.currentTime}
    test -d "$(dirname "$out")/../var/nix"
    touch $out
  ''
```

roberth reviewed (approved):
Not happy about the formatting at all, but I don't want to block the process of implementing formatting, because that would be even worse.
:+1: from me.

infinisil commented:
@roberth Even if the formatting isn't perfect, having a standard enforced formatting is better than the current churn. Considering that these cases you highlighted are somewhat rare, I don't think we need to block on this. We can also improve the formatter and then do another reformatting pass later (which will then be much smaller in diff size!). If you have more feedback about the formatting, it would be best to open issues in https://github.com/NixOS/nixfmt instead btw :)

roberth commented:
> Even if the formatting isn't perfect, having a standard enforced formatting is better

As you've read, I agreed with this part.
However, considering that this aspect of the formatter is still subject to change, it seems that you have the priority inversed:

> than the current churn.

Unnecessary formatter-induced churn is not ok. If the formatter doesn't behave correctly, or some aspects of it are still in development, don't both users with it.
I'm glad we've found this issue early, so that we can resolve it before bothering others with bad layouts and unnecessary churn.

> I don't think we need to block on this.

So I think we should block on this, fix the formatter, and do another pass.



> Considering that these cases you highlighted are somewhat rare

I assume you haven't run nixfmt on the whole repo.
There's two of them in this small PR, and I would expect more `runCommand` calls to be affected, because often it's just a single  `nativeBuildInputs` for simple tests.
But you're right, we should be doing more testing ;)

infinisil commented:
> If the formatter doesn't behave correctly, or some aspects of it are still in development

I'd say we should distinguish between different levels of correctness:
1. Semantic correctness: That the AST isn't changed and that the formatter is idempotent. We have this verified on all Nixpkgs files [since recently](https://github.com/NixOS/nixfmt/pull/215) (and more recently also with `nix-instantiate --parse`).
2. Standard matching: That the formatter adheres to the [established standard](https://github.com/NixOS/nixfmt/blob/master/standard.md). This can be very fuzzy in certain edge cases, but as we progressed at establishing the initial standard in https://github.com/NixOS/rfcs/pull/166, @piegamesde did a great job keeping the implementation to match decently closely.
3. Standard correctness: That the standard doesn't have any inconsistencies in itself. We also tried this during RFC development.

So we have 1 now but probably not 2 or 3. These can arguably never be perfect and take more brainpower to resolve, so it's important to consider the cost/value tradeoff. (And in a twisted way, forcing something imperfect onto people might actually get us more help with improving it :P).

But okay, let's reconsider this and either improve this case or have a better justification for not needing to improve it now.

> I assume you haven't run nixfmt on the whole repo.

Complete opposite! We've been automatically running nixfmt on the entirety of Nixpkgs already during development of the RFC to help us decide (e.g. see [here](https://github.com/NixOS/rfcs/pull/166#discussion_r1481391155)), and we're continuing to run it on all PRs that change nixfmt (e.g. see [here](https://github.com/NixOS/nixfmt/pull/222#issuecomment-2232164409)), and of course there's also the [treewide reformatting PR](https://github.com/NixOS/nixpkgs/pull/322537) that is being worked on recently :wink:

roberth commented:
> > I assume you haven't run nixfmt on the whole repo.
> 
> Complete opposite!

Ah sorry, I was unclear. I meant to refer to this PR specifically, which I conclude is not tree-wide.



> > doesn't behave correctly

Oh, that was a bit harsh maybe, out of context at least. I meant to refer to your judgement that this may not be the desired behavior.

Anyway, I think we're on the same page now, and I'd really like to thank you for the effort you're putting into this!

philiptaron commented:
@infinisil, any reason not to merge this?

infinisil commented:
@roberth's feedback in https://github.com/NixOS/nixpkgs/pull/326430#discussion_r1679889039 and https://github.com/NixOS/nixpkgs/pull/326430#discussion_r1679891836 is now addressed with https://github.com/NixOS/nixfmt/pull/224, force pushed to make use of it: [diff](https://github.com/NixOS/nixpkgs/compare/a0427beafeec8f52201536bb59203af8c8dda862..8f425c25adc365ccddab932cae96cdbed96c115b)
